File: TOMIMSV4/DAL/Mantenimientos/Producto/ProductoTipo/clsLnProducto_tipo.vb
Imports System.Data.SqlClient

Public Class clsLnProducto_tipo

    Public Shared Sub Cargar(ByRef oBeProducto_tipo As clsBeProducto_tipo, ByRef dr As DataRow)

        Try

            With oBeProducto_tipo
                .IdTipoProducto = IIf(IsDBNull(dr.Item("IdTipoProducto")), 0, dr.Item("IdTipoProducto"))
                .IdPropietario = IIf(IsDBNull(dr.Item("IdPropietario")), 0, dr.Item("IdPropietario"))
                .Codigo = IIf(IsDBNull(dr.Item("Codigo")), "", dr.Item("Codigo"))
                .NombreTipoProducto = IIf(IsDBNull(dr.Item("NombreTipoProducto")), "", dr.Item("NombreTipoProducto"))
                .Activo = IIf(IsDBNull(dr.Item("Activo")), False, dr.Item("Activo"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .User_mod = IIf(IsDBNull(dr.Item("user_mod")), "", dr.Item("user_mod"))
                .Fec_mod = IIf(IsDBNull(dr.Item("fec_mod")), Date.Now, dr.Item("fec_mod"))
            End With

        Catch ex As Exception
            Throw ex
        End Try

    End Sub

    Public Shared Function Insertar(ByRef oBeProducto_tipo As clsBeProducto_tipo, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            Ins.Init("producto_tipo")
            Ins.Add("idtipoproducto", "@idtipoproducto", DataType.Parametro)
            Ins.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Ins.Add("codigo", "@codigo", DataType.Parametro)
            Ins.Add("nombretipoproducto", "@nombretipoproducto", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("user_mod", "@user_mod", DataType.Parametro)
            Ins.Add("fec_mod", "@fec_mod", DataType.Parametro)

            Dim sp As String = Ins.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDTIPOPRODUCTO", oBeProducto_tipo.IdTipoProducto))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBeProducto_tipo.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@CODIGO", oBeProducto_tipo.Codigo))
            cmd.Parameters.Add(New SqlParameter("@NOMBRETIPOPRODUCTO", clsPublic.Quitar_Caracteres_No_Permitidos(oBeProducto_tipo.NombreTipoProducto)))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_tipo.Activo))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_tipo.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_tipo.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_tipo.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_tipo.Fec_mod))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then
                lTransaction.Commit()
            End If

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeProducto_tipo As clsBeProducto_tipo, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            Upd.Init("producto_tipo")
            Upd.Add("idtipoproducto", "@idtipoproducto", DataType.Parametro)
            Upd.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Upd.Add("codigo", "@codigo", DataType.Parametro)
            Upd.Add("nombretipoproducto", "@nombretipoproducto", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("user_mod", "@user_mod", DataType.Parametro)
            Upd.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Upd.Where("IdTipoProducto = @IdTipoProducto")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            cmd.CommandType = CommandType.Text


            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If
            cmd.Parameters.Add(New SqlParameter("@IDTIPOPRODUCTO", oBeProducto_tipo.IdTipoProducto))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBeProducto_tipo.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@CODIGO", oBeProducto_tipo.Codigo))
            cmd.Parameters.Add(New SqlParameter("@NOMBRETIPOPRODUCTO", clsPublic.Quitar_Caracteres_No_Permitidos(oBeProducto_tipo.NombreTipoProducto)))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_tipo.Activo))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_tipo.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_tipo.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_tipo.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_tipo.Fec_mod))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            If Not Es_Transaccion_Remota Then
                lTransaction.Commit()
            End If

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Shared Function Eliminar(ByRef oBeProducto_tipo As clsBeProducto_tipo, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing
        Dim cmd As New SqlCommand()

        Try

            cmd.CommandType = CommandType.Text

            Dim sp As String = " Delete from Producto_tipo" &
             "  Where(IdTipoProducto = @IdTipoProducto)"

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If


            cmd.Parameters.Add(New SqlParameter("@IDTIPOPRODUCTO", oBeProducto_tipo.IdTipoProducto))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
            cmd.Dispose()
        End Try

    End Function

    Public Function Listar() As DataTable

        Try

            Const sp As String = "SELECT * FROM Producto_tipo"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            dad.Fill(dt)

            Return dt

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Obtener(ByRef oBeProducto_tipo As clsBeProducto_tipo) As Boolean

        Try

            Dim sp As String = "SELECT * FROM Producto_tipo" &
            " Where(IdTipoProducto = @IdTipoProducto)"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDTIPOPRODUCTO", oBeProducto_tipo.IdTipoProducto))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(oBeProducto_tipo, dt.Rows(0))
            End If

            Return True

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Get_Single_By_Nombre_By_IdPropietario(ByVal pNombre As String, ByVal pIdPropietario As Integer, pConnection As SqlConnection, pTransaction As SqlTransaction) As clsBeProducto_tipo

        Get_Single_By_Nombre_By_IdPropietario = Nothing

        Try

            Dim IdxProductoClasificacion As Integer = 0

            Dim vSQL As String = "SELECT TOP 1 * FROM producto_tipo 
                                  WHERE IdPropietario=@IdPropietario NombreTipoProducto like '%" & pNombre & "%'"

            Using lDTA As New SqlDataAdapter(vSQL, pConnection)

                lDTA.SelectCommand.CommandType = CommandType.Text
                lDTA.SelectCommand.Parameters.Add(New SqlParameter("@IdPropietario", pIdPropietario))
                lDTA.SelectCommand.Transaction = pTransaction

                Dim lDT As New DataTable()
                lDTA.Fill(lDT)

                If lDT IsNot Nothing AndAlso lDT.Rows.Count > 0 Then

                    Dim lRow As DataRow = lDT.Rows(0)
                    Dim Obj As New clsBeProducto_tipo()
                    Cargar(Obj, lRow)
                    Get_Single_By_Nombre_By_IdPropietario = Obj
                    Return Get_Single_By_Nombre_By_IdPropietario

                End If

            End Using

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Get_Single_By_IdTipoProducto(ByVal pIdTipoProducto As String, pConnection As SqlConnection, pTransaction As SqlTransaction) As clsBeProducto_tipo

        Get_Single_By_IdTipoProducto = Nothing

        Try

            Dim IdxProductoClasificacion As Integer = 0

            Dim vSQL As String = "SELECT TOP 1 * FROM producto_tipo WHERE IdTipoProducto = @IdTipoProducto "

            Using lDTA As New SqlDataAdapter(vSQL, pConnection)

                lDTA.SelectCommand.CommandType = CommandType.Text
                lDTA.SelectCommand.Transaction = pTransaction
                lDTA.SelectCommand.Parameters.AddWithValue("@IdTipoProducto", pIdTipoProducto)

                Dim lDT As New DataTable()
                lDTA.Fill(lDT)

                If lDT IsNot Nothing AndAlso lDT.Rows.Count > 0 Then

                    Dim lRow As DataRow = lDT.Rows(0)
                    Dim Obj As New clsBeProducto_tipo()
                    Cargar(Obj, lRow)
                    Return Obj

                End If

            End Using

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Get_Single_By_Codigo(ByVal pNombre As String, pConnection As SqlConnection, pTransaction As SqlTransaction) As clsBeProducto_tipo

        Get_Single_By_Codigo = Nothing

        Try

            Dim IdxProductoClasificacion As Integer = 0

            Dim vSQL As String = "SELECT TOP 1 * FROM producto_tipo WHERE Codigo = '" & pNombre & "'"

            Using lDTA As New SqlDataAdapter(vSQL, pConnection)

                lDTA.SelectCommand.CommandType = CommandType.Text
                lDTA.SelectCommand.Transaction = pTransaction

                Dim lDT As New DataTable()
                lDTA.Fill(lDT)

                If lDT IsNot Nothing AndAlso lDT.Rows.Count > 0 Then

                    Dim lRow As DataRow = lDT.Rows(0)
                    Dim BeTipoProducto As New clsBeProducto_tipo()
                    Cargar(BeTipoProducto, lRow)
                    Return BeTipoProducto

                End If

            End Using

        Catch ex As Exception
            Throw ex
        End Try

    End Function

End Class

File: WMS.DALCore/Producto/clsLnProducto_tipo.cs
using Microsoft.Data.SqlClient;
using System.Data;
using System.Diagnostics;
using System.Reflection;
using WMS.EntityCore.Producto;
using Microsoft.Extensions.Configuration;
public class clsLnProducto_tipo
{

    private static clsInsert Ins = new clsInsert();
    private static clsUpdate Upd = new clsUpdate();

    public static void Cargar(ref clsBeProducto_tipo oBeProducto_tipo, DataRow dr)
    {
        int GetInt(string col) { return dr[col] is DBNull ? 0 : Convert.ToInt32(dr[col]); }
        bool GetBool(string col) { return dr[col] is DBNull ? false : Convert.ToBoolean(dr[col]); }
        string GetString(string col) { return dr[col] is DBNull ? "" : (Convert.ToString(dr[col]) ?? ""); }
        DateTime GetDate(string col) { return dr[col] is DBNull ? DateTime.Now : Convert.ToDateTime(dr[col]); }

        try
        {
            oBeProducto_tipo.IdTipoProducto = GetInt("IdTipoProducto");
            oBeProducto_tipo.IdPropietario = GetInt("IdPropietario");
            oBeProducto_tipo.NombreTipoProducto = GetString("NombreTipoProducto");
            oBeProducto_tipo.Activo = GetBool("Activo");
            oBeProducto_tipo.User_agr = GetString("user_agr");
            oBeProducto_tipo.Fec_agr = GetDate("fec_agr");
            oBeProducto_tipo.User_mod = GetString("user_mod");
            oBeProducto_tipo.Fec_mod = GetDate("fec_mod");
            oBeProducto_tipo.Codigo = GetString("codigo");
        }
        catch (Exception ex)
        {
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = sf?.GetMethod();
            string vMsgError = string.Format("{{0}} {{1}}", currentMethodName, ex.Message);

            throw new Exception(vMsgError);
        }
    }

    public static int Insertar(clsBeProducto_tipo oBeProducto_tipo, IConfiguration config, SqlConnection? cn = null, SqlTransaction? tx = null)
    {
        SqlConnection? cnLocal = null;
        SqlTransaction? txLocal = null;
        bool remota = cn != null && tx != null;

        try
        {
            Ins.Init("producto_tipo");
            Ins.Add("idtipoproducto", "@idtipoproducto", "F");
            Ins.Add("idpropietario", "@idpropietario", "F");
            Ins.Add("nombretipoproducto", "@nombretipoproducto", "F");
            Ins.Add("activo", "@activo", "F");
            Ins.Add("user_agr", "@user_agr", "F");
            Ins.Add("fec_agr", "@fec_agr", "F");
            Ins.Add("user_mod", "@user_mod", "F");
            Ins.Add("fec_mod", "@fec_mod", "F");
            Ins.Add("codigo", "@codigo", "F");

            string sql = Ins.SQL();

            if (!remota)
            {
                cnLocal = new SqlConnection(config.GetConnectionString("CST"));
                cnLocal.Open();
                txLocal = cnLocal.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            using var cmd = new SqlCommand(sql, remota ? cn! : cnLocal!, remota ? tx! : txLocal!);
            Bind(cmd, oBeProducto_tipo);
            int result = cmd.ExecuteNonQuery();

            if (!remota)
                txLocal!.Commit();

            return result;
        }
        catch (Exception ex)
        {
            if (!remota && txLocal is not null)
                txLocal.Rollback();

            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (!remota)
            {
                txLocal?.Dispose();
                if (cnLocal?.State == ConnectionState.Open)
                    cnLocal.Close();
                cnLocal?.Dispose();
            }
        }
    }

    public static int Insertar(IConfiguration config, clsBeProducto_tipo oBeProducto_tipo)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            Ins.Init("producto_tipo");
            Ins.Add("idtipoproducto", "@idtipoproducto", "F");
            Ins.Add("idpropietario", "@idpropietario", "F");
            Ins.Add("nombretipoproducto", "@nombretipoproducto", "F");
            Ins.Add("activo", "@activo", "F");
            Ins.Add("user_agr", "@user_agr", "F");
            Ins.Add("fec_agr", "@fec_agr", "F");
            Ins.Add("user_mod", "@user_mod", "F");
            Ins.Add("fec_mod", "@fec_mod", "F");
            Ins.Add("codigo", "@codigo", "F");

            string sql = Ins.SQL();
            using var cn = new SqlConnection(config.GetConnectionString("CST"));
            cn.Open();
            using var tx = cn.BeginTransaction();
            using var cmd = new SqlCommand(sql, cn, tx);
            Bind(cmd, oBeProducto_tipo);
            int result = cmd.ExecuteNonQuery();
            tx.Commit();
            return result;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);

            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
    }

    public static int Actualizar(clsBeProducto_tipo oBeProducto_tipo, IConfiguration config, SqlConnection? cn = null, SqlTransaction? tx = null)
    {
        Upd.Init("producto_tipo");
        Upd.Add("idtipoproducto", "@idtipoproducto", "F");
        Upd.Add("idpropietario", "@idpropietario", "F");
        Upd.Add("nombretipoproducto", "@nombretipoproducto", "F");
        Upd.Add("activo", "@activo", "F");
        Upd.Add("user_agr", "@user_agr", "F");
        Upd.Add("fec_agr", "@fec_agr", "F");
        Upd.Add("user_mod", "@user_mod", "F");
        Upd.Add("fec_mod", "@fec_mod", "F");
        Upd.Add("codigo", "@codigo", "F");
        Upd.Where("IdTipoProducto = @IdTipoProducto");

        string sql = Upd.SQL();
        bool remota = cn != null && tx != null;

        using var cnLocal = remota ? null : new SqlConnection(config.GetConnectionString("CST"));
        if (!remota) cnLocal!.Open();
        using var txLocal = remota ? null : cnLocal!.BeginTransaction();

        using var cmd = new SqlCommand(sql, remota ? cn! : cnLocal!, remota ? tx! : txLocal!);
        Bind(cmd, oBeProducto_tipo);
        int result = cmd.ExecuteNonQuery();
        if (!remota) txLocal!.Commit();

        return result;
    }
    public int Eliminar(IConfiguration config, clsBeProducto_tipo oBeProducto_tipo, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            const string sp = (" Delete from Producto_tipo" +
             "  Where(IdTipoProducto = @IdTipoProducto)");

            bool Es_Transaccion_Remota = (pConection != null && pTransaction != null);

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            cmd.Parameters.Add(new SqlParameter("@IdTipoProducto", oBeProducto_tipo.IdTipoProducto));

            int rowsAffected = cmd.ExecuteNonQuery();

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();

            return rowsAffected;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);

            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
    }

    public DataTable Listar(IConfiguration config)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            const string sp = "Select * FROM Producto_tipo";
            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
            SqlCommand cmd = new SqlCommand(sp, lConnection, lTransaction) { CommandType = CommandType.Text };
            SqlDataAdapter dad = new SqlDataAdapter(cmd);
            DataTable dt = new DataTable();
            dad.Fill(dt);

            lTransaction.Commit();

            return dt;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);

            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
    }

    public static bool GetSingle(IConfiguration config, ref clsBeProducto_tipo pBeProducto_tipo)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {

            const string sp = "Select * FROM Producto_tipo" +
            " Where(IdTipoProducto = @IdTipoProducto)";

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);

            SqlCommand cmd = new SqlCommand(sp, lConnection, lTransaction) { CommandType = CommandType.Text };
            SqlDataAdapter dad = new SqlDataAdapter(cmd);

            dad.SelectCommand.Parameters.Add(new SqlParameter("@IdTipoProducto", pBeProducto_tipo.IdTipoProducto));

            DataTable dt = new DataTable();
            dad.Fill(dt);

            lTransaction.Commit();

            if (dt.Rows.Count == 1)
            {
                DataRow r;
                r = dt.Rows[0];
                Cargar(ref pBeProducto_tipo, r);
                return true;
            }

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);

            throw new Exception(vMsgError);
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return false;

    }

    public static List<clsBeProducto_tipo> GetAll(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;
        List<clsBeProducto_tipo> lreturnList = new List<clsBeProducto_tipo>();

        try
        {
            const string sp = "Select * FROM Producto_tipo";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {

                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlDataAdapter lDTA = new SqlDataAdapter(sp, lConnection))
                    {
                        lDTA.SelectCommand.CommandType = CommandType.Text;
                        lDTA.SelectCommand.Transaction = lTransaction;
                        DataTable lDataTable = new DataTable();
                        lDTA.Fill(lDataTable);

                        clsBeProducto_tipo vBeProducto_tipo = new clsBeProducto_tipo();

                        foreach (DataRow dr in lDataTable.Rows)
                        {
                            vBeProducto_tipo = new clsBeProducto_tipo();
                            Cargar(ref vBeProducto_tipo, dr);
                            lreturnList.Add(vBeProducto_tipo);
                        }

                        lTransaction.Commit();
                    }

                    lConnection.Close();

                }

            }

            return lreturnList;

        }
        catch (SqlException ex1)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            var st = new StackTrace();
            var sf = st.GetFrame(0);
            MethodBase? currentMethodName = null;
            if (sf != null) { currentMethodName = sf.GetMethod(); }
            string vMsgError = string.Format("{0} {1}", currentMethodName, ex1.Message);

            throw new Exception(vMsgError);
        }
    }

    public static int MaxID(IConfiguration config)
    {
        const string sql = "SELECT ISNULL(MAX(IdTipoProducto), 0) FROM producto_tipo";

        using var cn = new SqlConnection(config.GetConnectionString("CST"));
        cn.Open();

        SqlTransaction? tx = cn.BeginTransaction(IsolationLevel.ReadUncommitted);

        try
        {
            using var cmd = new SqlCommand(sql, cn, tx);
            int result = Convert.ToInt32(cmd.ExecuteScalar());
            tx.Commit();
            return result;
        }
        catch (Exception ex)
        {
            tx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int MaxID(IConfiguration config, SqlConnection? cn = null, SqlTransaction? tx = null)
    {
        const string sql = "SELECT ISNULL(MAX(IdTipoProducto), 0) FROM producto_tipo";
        bool remota = cn != null && tx != null;

        using var cnLocal = remota ? null : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? txLocal = null;

        try
        {
            if (!remota)
            {
                cnLocal!.Open();
                txLocal = cnLocal.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            using var cmd = new SqlCommand(sql, remota ? cn! : cnLocal!, remota ? tx! : txLocal!);
            int result = Convert.ToInt32(cmd.ExecuteScalar());

            if (!remota) txLocal?.Commit();
            return result;
        }
        catch (Exception ex)
        {
            if (!remota) txLocal?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static void Bind(SqlCommand cmd, clsBeProducto_tipo e)
    {
        cmd.Parameters.AddWithValue("@IdTipoProducto", e.IdTipoProducto);
        cmd.Parameters.AddWithValue("@IdPropietario", e.IdPropietario);
        cmd.Parameters.AddWithValue("@NombreTipoProducto", e.NombreTipoProducto);
        cmd.Parameters.AddWithValue("@Activo", e.Activo);
        cmd.Parameters.AddWithValue("@user_agr", e.User_agr);
        cmd.Parameters.AddWithValue("@fec_agr", e.Fec_agr);
        cmd.Parameters.AddWithValue("@user_mod", e.User_mod);
        cmd.Parameters.AddWithValue("@fec_mod", e.Fec_mod);
        cmd.Parameters.AddWithValue("@codigo", e.Codigo);
    }
    public static bool Existe(clsBeProducto_tipo oBe, IConfiguration config)
    {
        const string sql = "SELECT COUNT(1) FROM producto_tipo WHERE IdTipoProducto = @IdTipoProducto";

        try
        {
            using var cn = new SqlConnection(config.GetConnectionString("CST"));
            using var cmd = new SqlCommand(sql, cn);
            cmd.Parameters.AddWithValue("@IdTipoProducto", oBe.IdTipoProducto);

            cn.Open();
            return Convert.ToInt32(cmd.ExecuteScalar()) > 0;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int InsertOrUpdate(IConfiguration config, clsBeProducto_tipo oBe, SqlConnection? cn = null, SqlTransaction? tx = null)
    {
        return Existe(oBe, config)
            ? Actualizar(oBe, config, cn, tx)
            : Insertar(oBe, config, cn, tx);
    }

    public static bool Existe_By_Codigo(string Codigo, ref clsBeProducto_tipo pBeProductoTipo, SqlConnection cn, SqlTransaction? tx = null)
    {
        try
        {
            const string sql = "SELECT TOP 1 * FROM producto_tipo WHERE codigo = @codigo";

            using var cmd = new SqlCommand(sql, cn, tx);
            cmd.Parameters.AddWithValue("@codigo", Codigo);

            using var da = new SqlDataAdapter(cmd);
            var dt = new DataTable();
            da.Fill(dt);

            if (dt.Rows.Count == 1)
            {
                Cargar(ref pBeProductoTipo, dt.Rows[0]);
                return true;
            }

            return false;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }

    public static void Valida_Atributos(IConfiguration config, clsBeProducto_tipoMi3 entity, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        bool isExternalTx = conn != null && tx != null;
        var connection = isExternalTx ? conn! : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? localTx = null;

        try
        {
            if (!isExternalTx)
            {
                connection.Open();
                localTx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            var BeProductoTipo = new clsBeProducto_tipo();
            bool existe = Existe_By_Codigo(entity.Codigo, ref BeProductoTipo, connection, isExternalTx ? tx! : localTx!);

            if (!existe)
            {

                if (!string.IsNullOrEmpty(entity.Codigo))
                {
                    var BeTipoProducto = new clsBeProducto_tipo();
                    BeTipoProducto.IdTipoProducto= MaxID(config, connection, isExternalTx ? tx : localTx) + 1;
                    BeTipoProducto.IdPropietario = entity.IdPropietario;
                    BeTipoProducto.Codigo = entity.Codigo;
                    BeTipoProducto.NombreTipoProducto = entity.NombreTipoProducto ?? entity.Codigo;
                    BeTipoProducto.User_agr = "1";
                    BeTipoProducto.User_mod = "1";
                    BeTipoProducto.Fec_agr = DateTime.Now;
                    BeTipoProducto.Fec_mod = DateTime.Now;
                    BeTipoProducto.Activo = entity.Activo;
                    Insertar( BeTipoProducto, config, connection, isExternalTx ? tx : localTx);
                }

            }
            else
            {
                BeProductoTipo.Codigo = entity.Codigo;
                BeProductoTipo.NombreTipoProducto = entity.NombreTipoProducto ?? entity.Codigo;
                BeProductoTipo.User_mod = "1";
                BeProductoTipo.Fec_mod = DateTime.Now;
                BeProductoTipo.Activo = entity.Activo;
                Actualizar( BeProductoTipo, config, connection, isExternalTx ? tx : localTx);

            }

        }
        catch (SqlException ex)
        {
            if (!isExternalTx && localTx is not null)
                localTx.Rollback();

            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name}: {ex.Message}", ex);
        }
        finally
        {
            if (!isExternalTx)
            {
                connection.Close();
                connection.Dispose();
                localTx?.Dispose();
            }
        }
    }

}